File: app/src/main/java/com/example/qrscan/LoginActivity.java
package com.example.qrscan;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.example.qrscan.Retrofit.MyService;
import com.example.qrscan.Retrofit.RetrofitClient;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;

import java.util.Objects;

import io.reactivex.disposables.CompositeDisposable;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class LoginActivity extends AppCompatActivity {

    TextView createAccountTextView;
    TextInputEditText editTextPassword, editTextRegNumber;
    Button buttonSignIn;
    SharedPreferences sharedPreferences;
    String username;
    String password;

    CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Override
    protected void onStop() {
        compositeDisposable.clear();
        super.onStop();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // init prefs
        sharedPreferences = getSharedPreferences("LoggedInUserPrefs", Context.MODE_PRIVATE);

        // init view
        editTextRegNumber = findViewById(R.id.text_input_username);
        editTextPassword = findViewById(R.id.text_input_password);
        buttonSignIn = findViewById(R.id.button_sign_in);
        buttonSignIn.setOnClickListener(v -> {
            username = Objects.requireNonNull(editTextRegNumber.getText()).toString();
            password = Objects.requireNonNull(editTextPassword.getText()).toString();
            loginUser(username, password);

            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString("username", username);
            editor.putString("password", password);
            editor.apply();
        });

        createAccountTextView = findViewById(R.id.text_create_account);
        createAccountTextView.setOnClickListener(v -> {
            View register_layout = LayoutInflater.from(LoginActivity.this)
                    .inflate(R.layout.register_layout, null);

            new MaterialAlertDialogBuilder(LoginActivity.this)
                    .setIcon(R.drawable.ic_user)
                    .setTitle("Register account")
                    .setMessage("Please fill out all fields")
                    .setView(register_layout)
                    .setNegativeButton("CANCEL", (dialog, which) -> dialog.dismiss())
                    .setPositiveButton("REGISTER", (dialog, which) -> {
                        TextInputEditText editUsernameRegister = register_layout.findViewById(R.id.text_input_username_register);
                        TextInputEditText editPasswordRegister = register_layout.findViewById(R.id.text_input_password_register);

                        if (TextUtils.isEmpty(Objects.requireNonNull(editUsernameRegister.getText()).toString())) {
                            Toast.makeText(LoginActivity.this, "Username cannot be empty", Toast.LENGTH_SHORT).show();
                            return;
                        }

                        if (TextUtils.isEmpty(Objects.requireNonNull(editPasswordRegister.getText()).toString())) {
                            Toast.makeText(LoginActivity.this, "Password cannot be empty", Toast.LENGTH_SHORT).show();
                            return;
                        }

                        registerUser(editUsernameRegister.getText().toString(),
                                editPasswordRegister.getText().toString());
                    }).show();
        });
    }

    Retrofit retrofitClient = RetrofitClient.getInstance();
    MyService myService = retrofitClient.create(MyService.class);

    private void registerUser(String username, String password) {
        Call<ResponseBody> call = myService.registerUser(username, password);
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
                Toast.makeText(LoginActivity.this, "code: " + response.code(), Toast.LENGTH_LONG).show();
                Log.d("response", response.message());
                Log.d("response", String.valueOf(response.isSuccessful()));
            }

            @Override
            public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
                Toast.makeText(LoginActivity.this, "" + t, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void loginUser(String username, String password) {
        if (TextUtils.isEmpty(username)) {
            Toast.makeText(this, "Username cannot be empty", Toast.LENGTH_SHORT).show();
            return;
        }

        if (TextUtils.isEmpty(password)) {
            Toast.makeText(this, "Password cannot be empty", Toast.LENGTH_SHORT).show();
            return;
        }

        Call<ResponseBody> call = myService.loginUser(username, password);
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
                if (response.isSuccessful()) {
                    Toast.makeText(LoginActivity.this, "sign in successful", Toast.LENGTH_SHORT).show();
                    startActivity(new Intent(LoginActivity.this, MainActivity.class));
                } else if (response.code() == 401) {
                    Toast.makeText(LoginActivity.this, "invalid credentials", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
                Toast.makeText(LoginActivity.this, "" + t, Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/qrscan/LoginActivity.kt
package com.example.qrscan

import android.content.Intent
import android.os.Bundle
import android.provider.AlarmClock.EXTRA_MESSAGE
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.textfield.TextInputEditText
import kotlinx.coroutines.*
import org.jsoup.Connection
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.util.*


class LoginActivity : AppCompatActivity() {


    private val USER_AGENT =
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/101.0.4951.64 Safari/537.36"
    private val loginFormUrl = "https://github.com/login"
    private val loginActionUrl = "https://github.com/session"

    private lateinit var usernameInput: TextInputEditText
    private lateinit var passwordInput: TextInputEditText
    private lateinit var homePage: Connection.Response


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        usernameInput = findViewById(R.id.text_input_username)
        passwordInput = findViewById(R.id.text_input_password)
        val loginButton: Button = findViewById(R.id.button_sign_in)

//        login()

        loginButton.setOnClickListener {
            initializeHomePage()

//            if (getResponseCode() == 302) {
//                openMainActivity()
//            } else {
//                Toast.makeText(this, "Hello", Toast.LENGTH_SHORT).show()
//            }
        }
    }

    private fun getPassword(): String {
        return passwordInput.text.toString()
    }

    private fun getUsername(): String {
        return usernameInput.text.toString()
    }


    // redirect to MainActivity
    private fun openMainActivity() {
        val intent = Intent(this, MainActivity::class.java).apply {
            putExtra(EXTRA_MESSAGE, getUsername() + getPassword())
        }
        startActivity(intent)
    }

    //  get login status code
//    private fun getResponseCode(): Int {
//        return homePage.statusCode()
//
//    }

//################################    NEW METHOD IMPLEMENTATION   ################################
    // initialize homepage
@OptIn(DelicateCoroutinesApi::class)
private fun initializeHomePage() = GlobalScope.launch(Dispatchers.IO) {
    homePage = Jsoup.connect(loginActionUrl)
        .cookies(getCookies())
        .data(populateLoginForm())
        .method(Connection.Method.POST)
        .userAgent(USER_AGENT)
        .execute()
    println(homePage.parse().html())
    }


//    // get authToken
    private fun getAuthToken(): String {
        val loginForm = loginForm()
        val loginDocument: Document = loginForm.parse()

        return java.lang.String.valueOf(
            Objects.requireNonNull(
                loginDocument.select("#login > div.auth-form-body.mt-3 > form > input[type=hidden]:nth-child(1)")
                    .first()?.attr("value"),
            )
        )
    }
//
//    // get cookies
private fun getCookies(): MutableMap<String, String> {
       return loginForm().cookies()
    }
//
//    // populate login form with data
private fun populateLoginForm(): HashMap<String, String> {
    val formData: HashMap<String, String> = HashMap()

    formData["commit"] = "Sign in"
        formData["login"] = getUsername()
        formData["password"] = getPassword()
        formData["authenticity_token"] = getAuthToken()

    return formData
    }
//
//
//    // get login form
    private fun loginForm(): Connection.Response {
        return Jsoup.connect(loginFormUrl).method(Connection.Method.GET).userAgent(USER_AGENT)
            .execute()
    }


}

